VAULT-34675 (#30243)
hashicorp/vault: website/content/api-docs/system/internal-counters.mdx

in the queried period. The `end_time` is the end of the final month of the queried
period.

> <Note title="Behavior Differences between Vault Enterprise and Community Edition">

Author: Moved it over!

> <Note title="Behavior Differences between Vault Enterprise and Community Edition">

Author: Added ### Sample response before that content!

The behavior of the `start_time` and `end_time` parameters in the Counters API differs based on the Vault edition.

> In **Vault Enterprise (ENT)**, these parameters are must align with the full billing period. In **Vault Community

Reviewer: ```suggestion
In **Vault Enterprise (ENT)**, these parameters are must align with the entire default billing period of 12 months. In **Vault Community
```

Edition (CE)**, the API supports flexible historical time ranges, but excludes the current month and any future dates. 
See below for details.

</Note>

<Tabs>

<Tab heading="ENT">

> In **Vault ENT**, `start_time` and `ent_time` are aligned to the start and end of a billing period, respectively:

Reviewer: ```suggestion
In **Vault ENT**, `start_time` and `ent_time` are adjusted to align with the to the start and end of a the corresponding billing period:
```

- The Counters API only returns **data for an entire billing period**.

> - By default, the API returns activity data for the **current billing period**.

Reviewer: ```suggestion
- By default, the API returns activity data for the start of the **current billing period** until the **current moment**.
```

> - If `start_time` is provided, the query is aligned to the **start of the billing period** to which that timestamp belongs.

Reviewer: ```suggestion
- If  the provided`start_time` is outside of the current billing period, the query is aligned to entire default billing period to which `start_time` belongs.
```

> - If only `end_time` is provided, the query is aligned to the **start of the billing period** to which the `end_time` belongs.

Reviewer: ```suggestion
- If only `end_time` is provided, the query is aligned to the entire default billing period to which the `end_time` belongs.
```

> - If `start_time` and `end_time` do **not** correspond to the start and end of the **same billing period**, Vault Enterprise:

Reviewer: ```suggestion
- If `start_time` and `end_time` do **not** correspond to the **same billing period**, Vault Enterprise:
```

>    - Billing period will be aligned date specified in `start_time`

Reviewer: ```suggestion
   - Billing period will be aligned to the date specified in `start_time`
```

>    - Billing period will be aligned to the date specified in `start_time`

Author: @divyaac doesn't this sentence seem a little confusing because in reality it is not the billing period that gets aligned, but rather the given start and end times. The current sentence seems to mean we adjust the billing periods 🤔

Author: Maybe we can change it to: the `start_time` gets aligned to the start of the billing period it belongs to?

Reviewer: Yes good point. Let's change that

>    - **Aligns** the `start_time` to the start of the billing period (using either `start_time` or `end_time` as reference).

Reviewer: ```suggestion
   - Billing period will be aligned date specified in `start_time`
```

>    - Billing period will be aligned to the date specified in `start_time`

Reviewer: ```suggestion
   - `start_time` will be set the billing period that corresponds to the date specified in `start_time`
```

>    - **Sets** the `end_time` to the **end of that billing period**.

Reviewer: ```suggestion
   - `end_time` will be adjusted to the **end of that billing period**.
```

period.

<Note title="Behavior Differences between Vault Enterprise and Community Edition">

The behavior of the `start_time` and `end_time` parameters in the Counters API differs based on the Vault edition.
In **Vault Enterprise (ENT)**, these parameters are restricted to aligning with full billing periods only. In **Vault Community
Edition (CE)**, the API supports flexible historical time ranges, but excludes the current month and any future dates. 
See below for details.

</Note>

<Tabs>

<Tab heading="ENT">

In **Vault ENT**, `start_time` and `ent_time` are aligned to the start and end of a billing period, respectively:

- The Counters API only returns **data for an entire billing period**.
- By default, the API returns activity data for the **current billing period**.
- If `start_time` is provided, the query is aligned to the **start of the billing period** to which that timestamp belongs.
- If only `end_time` is provided, the query is aligned to the **start of the billing period** to which the `end_time` belongs.
- If `start_time` and `end_time` do **not** correspond to the start and end of the **same billing period**, Vault Enterprise:

>    - **Aligns** the `start_time` to the start of the billing period (using either `start_time` or `end_time` as reference).

Author: No, not necessarily the `current` billing period. It aligns it to any billing period that the values correspond to, it can be a billing period from 2 or 3 years ago. It aligns to the `current` period, if the values belong to the current billing period. For example, if the billing start of the current billing period is March 1, 2025 and the given start time is Feb 1, 2025, the given start time becomes aligned to the last billing period, not `current`. Let me know if that helps!

>    - **Aligns** the `start_time` to the start of the billing period (using either `start_time` or `end_time` as reference).

Reviewer: What happens if the start and end date provided are of different billing periods? For example start date is two years before the current date and end date is in the current billing period?

>    - **Aligns** the `start_time` to the start of the billing period (using either `start_time` or `end_time` as reference).

Author: Great question. since these two belong to more than one period, the current setup chooses only one billing period identified by start time, so the value of start time takes precedence, thus returning only data from 2 years before in this example.

>    - **Aligns** the `start_time` to the start of the billing period (using either `start_time` or `end_time` as reference).

Author: This behavior is indicated in acceptance criteria:
```
If the start_time and the end_time are not the beginning of a billing cycle and the end is not the end of the same billing period

Set start_time to the start of the billing period corresponding to the start_time if specified. Else, determine the billing period using the specified end_time.

Set end_time to the end of the billing period.
```

>    - **Sets** the `end_time` to the **end of that billing period**.

Reviewer: End of the current billing period?

Author: Not necessarily, Divya, it aligns it to the end of any billing period it corresponds too. If the end time belongs to the last cycle or the cycle from 2 years ago, it sets it accordingly. It does not have to be the `current` billing period.

Author: I should just delete the line `- **Sets** the `end_time` to the **end of that billing period**.`?

Author: I think you changed this line to - `end_time` will be adjusted to the **end of that billing period**.

   - **Returns a warning**:

   ```json
   {
      "warning": "start_time and end_time parameters can only be used to specify the beginning or end of the same billing period. The values provided for these parameters are not supported and are ignored. Showing the data for the entire billing period corresponding to start_time. If start_time is not provided, the billing period is determined based on the end_time."
   }
   ```

### Sample request
Assume the start of the billing period is January 1, 2023 and current date is December 31, 2025.

```shell-session
$ curl \
   --header "X-Vault-Token: ..." \
   --request GET \
   http://127.0.0.1:8200/v1/sys/internal/counters/activity?end_time=2023-12-31T00%3A00%3A00Z&start_time=2023-03-01T00%3A00%3A00Z
```

If the start_time or end_time is within the billing period or has overlapping billing periods, the response looks like this:

### Sample response
<CodeBlockConfig hideClipboard>

```json
{
   "request_id":"26be5ab9-dcac-9237-ec12-269a8ca647d5",
   "lease_id":"",
   "renewable":false,
   "lease_duration":0,
   "data":{
      :

>       // data returned for that billing period

Reviewer: We can probably exclude this portion

- By default, the API returns results **until the end of the last complete month**.
- If `end_time` falls in the **current month or a future date**, Vault CE:
   - **Ignores** the unsupported value.
   - Returns data **only up to the end of the last month**.
   - **Returns a warning**:

      ```json
      {
         "warning": "end_time parameter can only be used to specify a date until the end of previous month. The value provided for this parameter was in the current month or in the future date and was therefore ignored. The response includes data until the end of the previous month."
      }
      ```

### Sample request
Assume the current date is September 25, 2025.

```shell-session
$ curl \
   --header "X-Vault-Token: ..." \
   --request GET \
   http://127.0.0.1:8200/v1/sys/internal/counters/activity?end_time=2025-10-01T00%3A00%3A00Z&start_time=2025-01-01T00%3A00%3A00Z
```

### Sample response

<CodeBlockConfig hideClipboard>

   ```json
{
   "request_id":"26be5ab9-dcac-9237-ec12-269a8ca647d5",
   "lease_id":"",
   "renewable":false,
   "lease_duration":0,
   "data":{
      :

>       // data returned for that billing period

Reviewer: Does this render in Vercel? If not we can just remove these comments

Author: No, it does not render, i will remove them, thanks for noting this!
